### PR TITLE
Updated to match Terraform beyond Layer0 guide

### PR DESCRIPTION
**Pull request by sesh-kebab**
Updated the directory structure to be:
 - /src/
 - /example-1
 - /example-2
 - /example-3
 - README.md

Examples 1-3 have been added to aid the supporting guide.

### CODE REVIEW

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#
+# modules/guestbook_service/main.tf
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: remove comment

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  }
+}
+
+resource "layer0_service" "guestbook" {
+  name          = "${var.service_name}"
+  environment   = "${var.layer0_environment_id}"
+  deploy        = "${layer0_deploy.guestbook.id}"
+  load_balancer = "${layer0_load_balancer.guestbook.id}"
+
+  #scale = 3
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: remove

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  }
+}
+
+provider "aws" {
+  access_key = "${var.access_key}"
+  secret_key = "${var.secret_key}"
+  region     = "${var.region}"
+}
+
+# note we are prefixing with the environment name also to ensure there won't be conflicts
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: remove comment

File: terraform-beyond-layer0/example-1/modules/guestbook_service/output.tf [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#
+# modules/guestbook_service/output.tf
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/modules/guestbook_service/output.tf]: I'll just comment here for all files - remove comment on top

File: terraform-beyond-layer0/example-1/main.tf [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+module "guestbook" {
+  source = "./modules/guestbook_service"
+
+  name                    = "guestbook1"
+  access_key              = "${var.access_key}"
+  secret_key              = "${var.secret_key}"
+  layer0_environment_id   = "${layer0_environment.demo.id}"
+  layer0_environment_name = "${layer0_environment.demo.name}_guestbook"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/main.tf]: remove the `layer0_environment_name` variable - it isn't being used

File: terraform-beyond-layer0/example-1/modules/guestbook_service/variables.tf [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+variable "secret_key" {}
+
+variable "region" {
+  default = "us-west-2"
+}
+
+variable "layer0_environment_id" {}
+
+variable "layer0_environment_name" {}
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/modules/guestbook_service/variables.tf]: remove the `layer0_environment_name` variable - it isn't being used

File: terraform-beyond-layer0/example-1/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#
+# outputs.tf
+#
+
+output "services" {
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/outputs.tf]: the name of this output should probably be `service_url`

File: terraform-beyond-layer0/example-2/main.tf [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  access_key = "${var.access_key}"
+  secret_key = "${var.secret_key}"
+  region     = "${var.region}"
+}
+
+# create one layer0 environment for each map value defined in variable 'environments'
+resource "layer0_environment" "demo" {
+  count = "${length(var.environments)}"
+
+  name = "${lookup(var.environments, count.index)}_demo"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/main.tf]: Instead of a `map`, let's use a `list` for the `environments` variable. 

As an example, you can see how they use interpolation with the `public_subnets` variable here: https://github.com/terraform-community-modules/tf_aws_vpc/blob/master/main.tf#L65

Comment (zpatrick) [reply terraform-beyond-layer0/example-2/main.tf]: [nit] also, appending `_demo` to the name of the environment seems odd to me... as a consumer I'd expect layer0 environments named `dev`, `staging`, `prod` to be created

File: terraform-beyond-layer0/example-2/main.tf [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+# create one loadbalancer for each map value defined in variable 'environments'
+resource "layer0_load_balancer" "guestbook" {
+  count = "${length(var.environments)}"
+
+  name        = "${lookup(var.environments, count.index)}_guestbook_lb"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/main.tf]: I would keep the `name` the same value. Seems redundant to need to do 
```
l0 loadbalancer get dev:dev_guestbook_db
l0 loadbalancer get prod:dev_guestbook_db
``` 
When, if they had the same names, they could do:
```
l0 loadbalancer get dev:guestbook-lb
l0 loadbalancer get prod:guestbook-lb
```

File: terraform-beyond-layer0/example-2/main.tf [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  environment = "${element(layer0_environment.demo.*.id, count.index)}"
+
+  port {
+    host_port      = 80
+    container_port = 80
+    protocol       = "http"
+  }
+}
+
+# create a datasource for each map value defined in variable 'environments' which
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/main.tf]: [nit] I had to re-read this a couple time to make sense of it. Perhaps we should "dumb it down":
```
Create a template for each environment. Each template will reference an environment-specific dynamodb table. 
```

File: terraform-beyond-layer0/example-2/main.tf [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+# - load balancer
+resource "layer0_service" "guestbook" {
+  count = "${length(var.environments)}"
+
+  name          = "${element(layer0_environment.demo.*.name, count.index)}_guestbook_svc"
+  environment   = "${element(layer0_environment.demo.*.id, count.index)}"
+  deploy        = "${element(layer0_deploy.guestbook.*.id, count.index)}"
+  load_balancer = "${element(layer0_load_balancer.guestbook.*.id, count.index)}"
+
+  # scale       = "${lookup(var.service_scale, lookup(var.environments, count.index), "1")}"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/main.tf]: remove or add

File: terraform-beyond-layer0/example-2/main.tf [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+# create a layer0 service for each map value defined in variable 'environments' which 
+# references an environment specific:
+# - layer0 environment
+# - deploy
+# - load balancer
+resource "layer0_service" "guestbook" {
+  count = "${length(var.environments)}"
+
+  name          = "${element(layer0_environment.demo.*.name, count.index)}_guestbook_svc"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/main.tf]: redundant naming (same as loadbalancer comments above)

File: terraform-beyond-layer0/example-2/output.tf [hunk 13]
```diff
@@ -0,0 +1,8 @@
+#
+# ./example2/output.tf
+#
+
+# Note the syntax for referencing resources using the count property
+output "guestbook_urls" {
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/output.tf]: this should be a `list` output, not a string. For reference, see: https://github.com/terraform-community-modules/tf_aws_vpc/blob/master/outputs.tf#L25

Comment (zpatrick) [inline terraform-beyond-layer0/example-2/output.tf]: [nit] I'd output the entity ids of each layer0 resource created:
```
output "environment_ids" {
  value = ["${layer0_environment.demo.*.id}"]
}

output "service_ids" {
  value = ["${layer0_service.guestbook.*.id}"]
}

```

File: terraform-beyond-layer0/example-2/variables.tf [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+# to add another environment, append the map variable with a new value
+# e.g., "2" = "prod"
+variable "environments" {
+  type = "map"
+
+  default = {
+    "0" = "dev"
+    "1" = "staging"
+
+    # "2" = "prod"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-2/variables.tf]: remove or add back in

File: terraform-beyond-layer0/example-3/env-dev/main.tf [hunk 15]
```diff
@@ -0,0 +1,27 @@
+#
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-3/env-dev/main.tf]: How do you feel about `example-3`? 

I know I've been going back and forth on this configuration, but having now looked at it, I feel like it's a lot of copied configuration with little value added. 

The only take-away I'm getting is how to keep terraform configurations separated from one another - something we could just tell the users in a small doc. 

The only other argument I can think for using this pattern would be if you needed different configurations for each environment - but I'd argue that can be achieved by adding variables and more interpolation to the modules (like how you do different scales for each environment in `example-2`).

**Review by zpatrick — CHANGES_REQUESTED** [review]

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf (hunk 1 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: ✅

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf (hunk 2 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: The guide references this and asks the user to uncomment the scale property to increase the scale. https://github.com/quintilesims/layer0/pull/184/files#diff-f672345b0c16fae2a231ebbed42c512cR403

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf (hunk 3 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: ✅

File: terraform-beyond-layer0/example-1/modules/guestbook_service/output.tf (hunk 4 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/output.tf]: ✅

File: terraform-beyond-layer0/example-1/main.tf (hunk 5 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/main.tf]: ✅

File: terraform-beyond-layer0/example-1/modules/guestbook_service/variables.tf (hunk 6 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/variables.tf]: ✅

File: terraform-beyond-layer0/example-1/outputs.tf (hunk 7 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/outputs.tf]: ✅

File: terraform-beyond-layer0/example-2/main.tf (hunk 8 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/main.tf]: map -> list ✅ 

re: [nit] not sure about this one. `_demo` suffix I anticipate would be for name of a application. I.e., if I had multiple applications, I anticipate environments like:

- dev_application1
- staging_application1
- dev_application2
- staging_application2

File: terraform-beyond-layer0/example-2/main.tf (hunk 9 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/main.tf]: oops missed this change - ✅

File: terraform-beyond-layer0/example-2/main.tf (hunk 10 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/main.tf]: ✅

File: terraform-beyond-layer0/example-2/main.tf (hunk 12 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/main.tf]: ✅

File: terraform-beyond-layer0/example-2/main.tf (hunk 11 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/main.tf]: @zpatrick this is also referenced in the guide to make it easier for the user who is walking through the guide to change the terraform file. not a fan?

File: terraform-beyond-layer0/example-2/output.tf (hunk 13 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/output.tf]: that makes it a lot easier. 👍

Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/output.tf]: ✅

File: terraform-beyond-layer0/example-2/variables.tf (hunk 14 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-2/variables.tf]: ✅

File: terraform-beyond-layer0/example-3/env-dev/main.tf (hunk 15 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-3/env-dev/main.tf]: One of the main take-aways from example 3 that I wanted to get across is how to structure your configuration files, such that you have environment specific state files.

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf (hunk 2 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf]: i see - perhaps comment above it then?

File: terraform-beyond-layer0/example-2/main.tf (hunk 8 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-2/main.tf]: While that architecture can definitely work, it somewhat undermines the concept of Layer0 environments. 
We've designed them as a shared space for multiple services to run. 

I think the preferred architecture we want to advertise is: 
*  dev
    * application1
    * application2
*  staging
    * application1
    * application2

Admittedly, it doesn't come across very well in this example since we only run one service in the environment. 
Curious to hear others' input @diemonster @tlake

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf [hunk 16]
```diff
@@ -1,9 +1,5 @@
-#
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf:1]: Do we have any walkthroughs on how to write a good terraform module for your service? 
When we first started using terraform with our services, we would just add a `layer0.tf` in the same directory as the service, kind of like a `Dockerfile`. Now we know that "releasing" it as a terraform module is a much better option. 

I assume most people would make the same mistake as us at first. I think it would be a good idea to steer them in the right direction here.

**Review by zpatrick — CHANGES_REQUESTED** [review]

File: terraform-beyond-layer0/example-2/main.tf (hunk 11 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-2/main.tf]: I like it, I'd vote we leave it in the demo without comments.

File: terraform-beyond-layer0/example-3/env-dev/main.tf (hunk 15 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-3/env-dev/main.tf]: I see what you mean - given this is more "simple" terraform config, it kind of makes more sense to me as example 2. Then, we compact everything down in example 3. 

To add to the value of this example, I think we should add add some sections talking about how to "release" your services as modules. I'd be open to pair on it.

File: terraform-beyond-layer0/example-3/env-dev/variables.tf [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  default = "dev"
+}
+
+variable "region" {
+  default = "us-west-2"
+}
+
+# rather than add conditional statements to resources configurations
+# creating a lookup table can be more useful
+variable "service_scale" {
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-3/env-dev/variables.tf]: I don't think this is used in example3

File: terraform-beyond-layer0/example-3/env-staging/variables.tf [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+variable "region" {
+  default = "us-west-2"
+}
+
+# rather than add conditional statements to resources configurations
+# creating a lookup table can be more useful
+variable "service_scale" {
+  type = "map"
+
+  default = {
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-3/env-staging/variables.tf]: ditto

File: terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json [hunk 19]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json:1]: i think for this (and perhaps example2), we should place the module in a child directory, or even with the source code for  the guestbook example

File: terraform-beyond-layer0/example-3/modules/guestbook_service/main.tf [hunk 20]
```diff
@@ -54,19 +23,20 @@ resource "layer0_load_balancer" "guestbook" {
 }
 
 resource "layer0_service" "guestbook" {
-  name          = "guestbook"
+  name          = "${var.environment}_${var.name}_guestbook_svc"
```
Comment (zpatrick) [inline terraform-beyond-layer0/example-3/modules/guestbook_service/main.tf]: possibly redundant naming here?

**Review by zpatrick — CHANGES_REQUESTED** [review]

File: terraform-beyond-layer0/example-3/env-dev/variables.tf (hunk 17 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-3/env-dev/variables.tf]: It is used in main.tf and provides a default that can be modified by the consumer
```
module "guestbook" {
  source = "./../modules/guestbook_service"
  ...
  service_scale = "${var.service_scale}"
}
```

File: terraform-beyond-layer0/example-3/env-staging/variables.tf (hunk 18 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-3/env-staging/variables.tf]: see above

File: terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json (hunk 19 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json:1]: @zpatrick not sure what you mean here? So the directory structure for `Dockerrun.aws.json` is

```
# current path
/example-3/modules/guestbook_service/Dockerrun.aws.json

# suggested path?
/example-3/modules/guestbook_service/src/Dockerrun.aws.json
```

File: terraform-beyond-layer0/example-3/env-dev/variables.tf (hunk 17 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-3/env-dev/variables.tf]: I see - however, I now see the issue is with the module design. This input should be an integer. As is, the module is tightly coupled with your scaling implementation. The map/scale interpolation should happen at this layer of config, not inside the module.

File: terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json (hunk 19 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json:1]: Sorry, that statement was pretty vague. 

This kind of feeds of my earlier comment about showing users how to write good Terraform modules. Basically, in the 'real world' equivalent that this example is trying to show, the module for the `guestbook` service you are importing wouldn't exist in the same repo/directory tree. The module should be a "blueprint" for a service, and the configuration should be the real, actual resources created - and its very important we keep these things separate. 

This is something I'm not seeing in these examples, and this example is good candidate for it. By putting the "blueprint" module for the `guestbook` service as a subdir in this example, we are showing a pattern users shouldn't do

Comment (sesh-kebab) [reply terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json:1]: I think I know what you're saying but not sure I completely agree. 

> the module for the guestbook service you are importing wouldn't exist in the same repo/directory tree

This might be the case if you intend for your modules to be reusable components across different projects but that doesn't have to be the case right? You can just use modules to group a set of resources or provide a layer of abstraction. In the case of example 3, the modules directory exists outside the the different environments. I.e.,

```
example-3
  |-env-dev
  |-env-staging
  |-modules
      |- guestbook_service
```

I would argue that the purpose here is to avoid duplicate configuration per environment (dev, staging). In theory, the env-dev and env-staging can import the guestbook_service module from  anywhere. Placing the guestbook_service in the same repo/directory tree is more for convenience but I don't know if we can conclusively say that it is a pattern users shouldn't use?

File: terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf (hunk 16 above)
Comment (sesh-kebab) [reply terraform-beyond-layer0/example-1/modules/guestbook_service/main.tf:1]: We don't have any walkthroughs that focus on writing good terraform modules I believe. At least, not to my knowledge. Might have to pair up on this somehow; need further elaboration on "releasing" terraform modules.

File: terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json (hunk 19 above)
Comment (zpatrick) [reply terraform-beyond-layer0/example-3/modules/guestbook_service/Dockerrun.aws.json:1]: > You can just use modules to group a set of resources or provide a layer of abstraction

Definitely - but `example-3/modules/guestbook_service` is not that.
This is the exact same trap we fell into when first starting w/ Terraform.
Your module here is mixing the "blueprint" config and the "real" config.
